README


README for OPIUM

        Copyright (C) 1998-2010 The OPIUM Group

	This file is part of OPIUM

        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2, or (at your option)
        any later version.

        This program is distributed in the hope that it will be useful,
        but WITHOUT ANY WARRANTY; without even the implied warranty of
        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        GNU General Public License for more details.

        You should have received a copy of the GNU General Public License
        along with this program; see the file COPYING.  If not, write to
        the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
        Boston, MA 02111-1307, USA.


This directory contains the full source of OPIUM, which is an open-source
pseudopotential generation program.

Please see the file COPYING for copying conditions.

Please see the files in the doc/ directory for more information.

See the file INSTALL for compilation and installation instructions,

Additional information can be found on the OPIUM website: http://opium.sourceforge.net




doc/CHANGELOG

Opium changelog


1.0.1 --> 1.0.2

	1) Further improvement of handling of s,p,d ordering problem
	   Tried the following tests for .pwf, .ncpp, and .fhi output styles:
	    Ag - 4d,5s,5p valence ordering s-local
	    Ag - 5s,5p,4d                  s-local
	    Ag - 4d,5s,5p                  local not specified (check default = s NOT 1st state)

	    Ti - 3s,3p,3d,4s,4p            s-local
	    Ti - 3s,4s,3p,4p,3d            s-local
	    Ti - 3s,3p,4s,4p,3d            '0' local (l=0)

	    Li - 1s,2s,2p                  local not specified
	    Li - 1s,2p,2s                  local not specified
	    
	    For each group, bh, abinit, and pwscf give the same results to ~8 digits.

	2)  Code should be totally case insensitive (all strcmp's replaced with strcasecmp).

	3)  Param file appened to fhi, ncpp files (thanks Wissam!)

	4)  Plot help 'vi' example corrected.

        5)  Added fclose() in do_nl.c (thanks Keith!)

        6)  Configure script modifed to allow LDFLAGS to be appended from command line

        7)  Fixed spelling of Schrodinger in descreen.f
 
        8)  AE and PS eigenvalues were actually reversed in descreen.f (they are
            identical so no big deal).

        9)  Fixed bug in nllogd for KB log-derivative (missed common block)
 
        10) Adjusted do_logplt so 'ae ps nl' need not be repeated to make plot (thanks Rongqin!)

        11) Added line for refernece energy in logder plot (thanks Rongqin!)

        12) Modified treatment of unbound states to work with semicore (thanks Sara!)

        13) Modified output for .ncpp and .fhi to indicate PW XC correctly

        14) Changed MacOSX configure defs so -r8 included by default

        15) Major modfication of ghost state reporting, now states whether ghost testing
            was inconclusive (no bound states for local potential)

        16) Ghost testing added for unbound states as local or non-local

        17) Interp.f output cleaned up

        18) Logder plotting states that no [loginfo] key block is set (not just a core dump!)

        19) Fixed legends for plots that were not correct

8/19/04 - EJW



1.0.2 --> 1.0.3

1) Ghost testing fixed, removed the '?' output if a local state was
not bound.  Now this eigenvalue is treated as if its eigenvalue = 0.0

2) Changed the way an unbound state is selected.  Now, if the filling
is set to <-1.0e-12, the state is treated as unbound (and treated as
if it is empty).  The eigenvalue is fixed by the guess or set to 0.0
if no guess specified.

3) Added the .recpot output format routines (contributed by Keith
Refson).  This format is native to the solid state code CASTEP.

4) Added GNU license information, license header prepended to every .f
.c and .h file and added a "COPYING" file which contains the full-text
of the license.

5) Fixed bug related to DNL in the PWF format (introduced in the
previous version).

6) Added the Fuchs + Scheffler partial core correction method which is
implemented in fhi98pp.

7) Fixed bug related to labeling the eigenvalues in the logdplot

8) Changed keyblock help to reflect #2 and #6 above.

9) Configure script changed such that ifort and ifc libcxa gets
statically linked (requested by Keith Refson).

10) Added true NULL characters to nwrite calls in writepwf for better
compatibility.

10/12/04 - EJW


1.0.3 --> 2.0.1

1) Troullier-Martins potential method added 

2) Cleaned up plotting output

3) Added calculation of the convergence of the kinetic energy error, output is sent to both rpt and log file

4) Can now plot the KE convergence

5) Explicit annoucement of the presnece of ghost states in the rpt file

6) Changed configure script to generate ONE set of makevars such that users can edit them if configure fails

7) Added listing of compile and execution platform in log file

8) Fixed some errors in the element symbol table

9) Increased number of keyblock reads so that keyblocks ending in string reads don't screw things up.

10) Added the ability to set an environmental variable called $XMGRACE_OPTS which is included in the command-line call to xmgrace.

11) Increased the dimensions of all parameters such that 5f elements can be constructed

9/10/05 - EJW

2.0.1 --> 2.0.2

1) Bug fix for ncpp (PWSCF) style output.  Wavefunctions had wrong number of r's in 1.0.2, 1.0.3, and 2.0.1.  See doc/README_solidstatecodes
for more information.


2.0.2 --> 2.0.3

1) Bug fix for f-projectors in .pwf output format

2) Applied patch from Keith Refson to repair .recpot output format
3) Some (but not all) sprintf statments changed to snprintf

4) BETA level implementation of .psf (siesta) format and .cpmd (CPMD) output format,
see 'opium -c' for more information

5) GGA functional allows for smoothing near the origin (see 'opium -k' for how to
invoke).

6) Corrections to KE convergence output when semicore states are involved.  Now the
KE convergence of all valence orbitals is computed.

2.0.3 --> 2.0.4

1) .pwf output file now contains the wavefunctions from the MOST RECENT configuration 
at the time of generation.  For example, 

> ./opium al log ae ps nl pwf 

will include the wavefunctions of the reference state but, 

> ./opium al log ae ps nl tc pwf 

will include the wavefunctions of the last test configuration.

This is done since often the reference is not the desired ground state atomic
configuration.

The log file also reports which configuration was used.

2) (finally) included the rudimentary emacs OPIUM mode file "_emacs_mode". This
can be appended to ones ~/.emacs init file to included highlighting of the
acceptable keyblocks.


2.0.4 --> 2.0.5

1) Bugfix for new .pwf output (eigenvalues match the configuration used for the
output wavefunctions).

2) Useful information (compile and execution host, os, date, version, 
and command line used for psp generation) written at the end of most psp
output file types.

3) Bugfix in do_ke.c and kcomp.F (dimensioning error)

4) Bugfix for 'all' command -- no longer attempts 'fc' step.


src/atom/hfde.F
c--------------------------------------------------------------------
c The following code is adapted from:
c
c A GENERAL HARTREE-FOCK PROGRAM
c
c     by C. Froese Fischer
c        Vanderbilt University
c        Nashville, TN 37235 USA
c
c     April, 1987
c     Published in Computer Physics Communications, 43 (1987) 355
c
c     .... Modified for for arbitrarily filled f-shells
c     by Gediminas GAIGALAS 
c        Institute of Theoretical Physics and Astronomy
c        A. Gostauto str. 12
c        Vilnius, 2600, Lithuania
c       
c     January, 1996
c
c     Pulished in Computer Physics Communications, 98 (1996) 255.
c
c
c
c
c Obtained from http://atoms.vuse.vanderbilt.edu/Elements/CompMeth/hf.f
c-----------------------------------------------------------------------


c
c Copyright (c) 1998-2012 The OPIUM Group
c
c This program is free software; you can redistribute it and/or modify
c it under the terms of the GNU General Public License as published by
c the Free Software Foundation; either version 2 of the License, or
c (at your option) any later version.
c
c This program is distributed in the hope that it will be useful,
c but WITHOUT ANY WARRANTY; without even the implied warranty of
c MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
c GNU General Public License for more details.
c
c You should have received a copy of the GNU General Public License
c along with this program; if not, write to the Free Software
c Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
c
c
      subroutine hfde(m,ipsp,zeff,dpm,iprint)
      implicit double precision (a-h,o-z)
#include "fortdim.h"

c -------------------------------------------------------------------------
c     External (shared between C and Fortran) common blocks
c -------------------------------------------------------------------------
      common /aorb/ ncore,nval,norb,nlm(n0),
     $     no(n0),lo(n0),nmax(n0),maxim
      common /adat/ wnl(n0),en(n0),so(n0),xion
      common /aval/ rcall(n0),rvap(n0),rnorm(n0),ibd(n0),etot
      common /wfn/ rnl(npdm,n0)
      common /consts/ etol,vtol,maxit,isoft
      common /grid/ h,r1,z,r(npdm),np
      common /edi/ e(n0,n0)
      common /ahf/ s(n0),az(n0),acc(n0)
      common /ipr/ ipr

c -------------------------------------------------------------------------
c     Internal (Fortran only) common blocks                
c -------------------------------------------------------------------------
      common /totpot/ rvcore(npdm,n0),rvps(npdm,n0),rvcoul(npdm)
      common /iterm/ iterm
      common /elim/ emore,eless,elim
      common /ipos/ ipos(n0),itermcount
      common /itoo/ ietoolo,ietoohi
      common /iskip/ iskip
      common /ameth/ meth(n0),imeth

      dimension p(npdm),dpm(n0),rsatom(npdm)
      
      logical fail,ilast

      ed1=0.d0
      ed2=e(m,m)
      imeth=max(1,meth(m))
      if (norb.eq.1) then
         meth(m)=2
         imeth=2
      endif

      call hfmethd(m,ipsp,p,zeff,azz,fail,iprint)

      if (fail) goto 25

 12   continue

      pnn=dsqrt(quad(m,nmax(m),p,p))
      do k=1,nmax(m)
         p(k)=p(k)/pnn
      enddo

c      write(7,*) 'azz 1',azz,pnn
      azz=azz/pnn

      if (imeth.eq.1) then
         if (abs(1.d0-ed2/e(m,m)).lt.0.005d0.and.
     $        dmax1(dabs(1.d0-pnn),dabs(1.d0/pnn - 1.d0)).
     $        gt.0.20d0) then
            meth(m)=2
            imeth=2
            goto 25
         endif
      endif

      if (ipr.ne.m) then
         acc(m)=0.75*acc(m)
      else
         ed2=ed2-e(m,m)
         if (ed1*ed2.gt.0) then
            acc(m)=0.75*acc(m)
         else
            acc(m)=(1.d0 + 3.d0*acc(m))/4.d0
         endif
      endif
      c=acc(m)
      cd=1.d0-c

      !using a fixed acc parameter
      !c = 0.5
      !cd = 1.0-c

c      write(7,*) 'ACC', m,acc(m),c,ipr

      if (acc(m).gt.0.99999) acc(m)=0.50

      !do mm=1,norb
      !  do i = 1,np
      !    write(31,*) rnl(i,mm)
      !  enddo
      !enddo


      dp=0.d0
      do j=1,nmax(m)
c      do j=1,np
         diff=rnl(j,m)-p(j)
         dp=max(dp,dabs(diff)*sqrt(r(j)))
         rnl(j,m)=p(j)+c*diff
      enddo
      do i=nmax(m)+1,np
         rnl(i,m)=0.d0
      enddo

      !different in 4th decimal
      !do mm=1,norb
      !  do i = 1,np
      !    write(32,*) rnl(i,mm)
      !  enddo
      !enddo


c      write(7,*) 'before de adj', m,az(m),azz

      az(m)=cd*azz+c*az(m)
      azz=az(m)

      !same
      !do mm=1,norb
      !   do i = 1,np
      !      write(30,*) rnl(i,mm)
      !   enddo
      !enddo



      if (wnl(m).lt.1e-6) then
         dpw=dp/0.001
      else
         dpw=dp/sqrt(wnl(m))
      endif

      !same
      !do mm=1,norb
      !  do i = 1,np
      !    write(30,*) rnl(i,mm)
      !  enddo
      !enddo



      do j=1,norb
         if (abs(e(m,j)).ne.0.d0.and.m.ne.j) then
            if (dpm(j).lt.dsqrt(wnl(j))*dpw) then
               nnp=max(nmax(j),nmax(m))
               rx=quadr(m,j,0)
c               write(7,63) nlm(j),nlm(m),rx
               do jj=1,nnp
                  rnl(jj,m)=rnl(jj,m)-rx*rnl(jj,j)
               enddo
               azz=azz-rx*az(j)
c               write(7,*) 'orth', j,azz,az(j),rx
            endif
         endif
      enddo

 63   format(6x,'<',i3,'|',i3,'>=',1pd8.1)      

      !same
      !do mm=1,norb
      !  do i = 1,np
      !    write(30,*) rnl(i,mm)
      !  enddo
      !enddo



      if (abs(rx).gt.0) then
         rx=dsqrt(quadr(m,m,0))
         
         do jj=1,np
            rnl(jj,m)=rnl(jj,m)/rx
         enddo
         azz=azz/rx
      endif

      !same
      !do mm=1,norb
      !  do i = 1,np
      !    write(30,*) rnl(i,mm)
      !  enddo
      !enddo



      jjset=np
 687  if (dabs(rnl(jjset,m)).lt.1e-15) then
         rnl(jjset,m)=0.d0
         jjset=jjset-1
         goto 687
      endif
      nmax(m)=jjset

      if (azz.gt.0.d0) az(m)=max(azz,az(m)/2.d0)

c      write(7,*) 'after de adj', m,az(m),azz

      en(m)=e(m,m)

      if (iprint.gt.1) 
     $     write (7,17) nlm(m),e(m,m),az(m),pnn,dp,r(nmax(m))
 17   format(20x,i3,2f15.7,f12.7,1pd10.2,1pd10.2)            

      maxim=max(maxim,nmax(m))
      dpm(m)=dp
      if(ipr.eq.m) then
         ed1=ed2
      else
         ed1=ed2-e(m,m) 
      endif
      ipr=m
      do i=1,nmax(m)
         rsatom(i)=rsatom(i)+wnl(m)*rnl(i,m)**2
      enddo

      !same
      !do mm=1,norb
      !  do i = 1,np
      !    write(30,*) rnl(i,mm)
      !  enddo
      !enddo



      return

 25   continue
      write(7,*) "Solve for HF wfn failed, orthogonalizing..."
      if (norb.eq.1) goto 27
      call orthog(ilast)
      call grange1
      
 27   continue
      call hfmethd(m,ipsp,p,zeff,azz,fail,iprint)
      if ( fail ) then
         if (imeth.eq.2) return
         imeth=2
         goto 27
      else
         goto 12
      endif
      
      return
      end
